bundles/org.eclipse.ui.workbench.texteditor/src/org/eclipse/ui/texteditor/FindReplaceOverlay.java

```diff
@@ -0,0 +1,995 @@
+				() -> findReplaceLogic.performSelectAll(getFindString()));
+	}
+
+	private KeyListener shortcuts = new KeyListener() {
+
+		private void performEnterAction(KeyEvent e) {
+			boolean isShiftPressed = (e.stateMask & SWT.SHIFT) != 0;
+			boolean isCtrlPressed = (e.stateMask & SWT.CTRL) != 0;
+			if (okayToUse(replaceBar) && replaceBar.isFocusControl()) {
+				if (isCtrlPressed) {
+					performReplaceAllOnEnter();
+				} else {
+					performReplaceOnEnter();
+				}
+			} else {
+				if (isCtrlPressed) {
+					performSearchAltOnEnter();
+				} else {
+					performSearchOnEnter(isShiftPressed);
+				}
+			}
+		}
+
+		private void performReplaceAllOnEnter() {
+			performReplaceAll();
+			evaluateFindReplaceStatus();
+		}
+
+		private void performReplaceOnEnter() {
+			performSingleReplace();
+			evaluateFindReplaceStatus();
+		}
+
+		private void performSearchAltOnEnter() {
+			performSelectAll();
+			evaluateFindReplaceStatus();
+		}
+
+		private void performSearchOnEnter(boolean isShiftPressed) {
+			performSearch(!isShiftPressed);
+			evaluateFindReplaceStatus();
+		}
+
+		@Override
+		public void keyPressed(KeyEvent e) {
+			e.doit = false;
+			if ((e.stateMask & SWT.CTRL) != 0 && (e.keyCode == 'F' || e.keyCode == 'f')) {
+				close();
+			} else if ((e.stateMask & SWT.CTRL) != 0 && (e.keyCode == 'R' || e.keyCode == 'r')) {
+				if (findReplaceLogic.getTarget().isEditable()) {
+					toggleReplace();
+				}
+			} else if ((e.stateMask & SWT.CTRL) != 0 && (e.keyCode == 'W' || e.keyCode == 'w')) {
+				toggleToolItem(wholeWordSearchButton);
+			} else if ((e.stateMask & SWT.CTRL) != 0 && (e.keyCode == 'P' || e.keyCode == 'p')) {
+				toggleToolItem(regexSearchButton);
+			} else if ((e.stateMask & SWT.CTRL) != 0 && (e.keyCode == 'A' || e.keyCode == 'a')) {
+				toggleToolItem(searchInSelectionButton);
+			} else if ((e.stateMask & SWT.CTRL) != 0 && (e.keyCode == 'C' || e.keyCode == 'c')) {
+				toggleToolItem(caseSensitiveSearchButton);
+			} else if (e.keyCode == SWT.CR || e.keyCode == SWT.KEYPAD_CR) {
+				performEnterAction(e);
+			} else if (e.keyCode == SWT.ESC) {
+				close();
+			} else {
+				e.doit = true;
+			}
+		}
+
+		private void toggleToolItem(ToolItem toolItem) {
+			toolItem.setSelection(!toolItem.getSelection());
+			toolItem.notifyListeners(SWT.Selection, null);
+		}
+
+		@Override
+		public void keyReleased(KeyEvent e) {
+			// Do nothing
+		}
+	};
+
+	private ControlListener shellMovementListener = new ControlListener() {
+		@Override
+		public void controlMoved(ControlEvent e) {
+			positionToPart();
+		}
+
+		@Override
+		public void controlResized(ControlEvent e) {
+			positionToPart();
+		}
+	};
+	private FocusListener overlayFocusListener = new FocusListener() {
+
+		@Override
+		public void focusGained(FocusEvent e) {
+			// do nothing
+		}
+
+		@Override
+		public void focusLost(FocusEvent e) {
+			findReplaceLogic.activate(SearchOptions.GLOBAL);
+			searchInSelectionButton.setSelection(false);
+		}
+
+	};
```

Comment:
Using the static listener factories provided by the typed listeners would simplify this any many of the listeners blow that only have one method implemented too.
```suggestion
	private FocusListener overlayFocusListener = FocusListener.focusLostAdapter(e -> {
		findReplaceLogic.activate(SearchOptions.GLOBAL);
		searchInSelectionButton.setSelection(false);
	});
```

Reply:
@HannesWell I didn't know these factories existed, but they make a lot of sense to use here! Would you consider this a blocker for this PR or is it fine if we create a follow-up-PR for this (imho very positive!) change?
I would prefer not to delay this PR for any longer as it has become quite big already.

Reply:
I'm glad I could share some knowledge. :)

I wouldn't consider it a blocker to not change it, but at the same time I don't think it is ideal to add something when there is the intention to remove it subsequently. Assuming that this is not controversial I would therefore appreciate it if you would apply it (but as said for me it's not a blocker).

I think the change is straight forward and it would make this PR smaller and thus a bit easier to understand.
And since the master is currently closed anyway and we are in the very beginning of a new release cycle I don't think there is a need to rush :)